### PR TITLE
Add CI build job barriers to prevent unnecessary builds

### LINKED ISSUE #3658: Make agent plugin tests in PR only necessary.
Today, we have 3 agent test jobs, lead by @dmsolr and jenkins setup by @kezhenxu94. With many contributors' help, especially @arugal, over 1/3 cases have been moved in. 
Even today, we have a little issue about job-3, it should be fixed by INFRA in the short term. And we already got a sponsor(tetrate.io)'s VMs(4) and @x22x22's one, plugin tests are still heavy payload. 
We all agree that, these jobs are significantly important for the project, but they are not necessary in all PR cases. Such as backend change only(no protocol change), UI change and documentation changed.

We should invest one test optimization starting by plugin tests, if and only if the `apm-sniffer` and `apm-protocol`(submodule) has changes, plugin tests should be done, otherwise, we could do a quick success status set for all 3 jobs.

### PR DESCRIPTION
**Pull request by kezhenxu94**
Please answer these questions before submitting pull request

- Why submit this pull request?
- [ ] Bug fix
- [x] New feature provided
- [ ] Improve performance

- Related issues

closes #3658 

Will test multiple times, label TBD for now

### CODE REVIEW

**Comment by wu-sheng** [comment]
My point is we need the reviewer to be notified no such ci processed in case some bug in change check or new module added. What do you think?

**Comment by kezhenxu94** [comment]
> My point is we need the reviewer to be notified no such ci processed in case some bug in change check or new module added. What do you think?

Make sense, in plan

File: tools/conditional/e2e-build-condition.sh [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+sh -x $(dirname "$0")/changed.sh --any-of '^pom.xml$'
+#  '^test/e2e/.*$' \
+#  '^apm-commons/.*$' \
+#  '^apm-protocol/.*$' \
+#  '^apm-sniffer/.*$' \
+#  '^oap-server/.*$' \
```
Comment (wu-sheng) [inline tools/conditional/e2e-build-condition.sh]: e2e should include webapp changes.

**Comment by wu-sheng** [comment]
One question, how do you test this? Remove one condition on purpose? Clearly, the ci should run as you changed the ci control file.

**Comment by wu-sheng** [comment]
~400/700+ test cases have been moved. Seems the priority of this PR is increasing :P 

@arugal is super crazy about shipping the test cases.

**Comment by kezhenxu94** [comment]
@wu-sheng the effect should look like this, if it looks good to you, I'll start to fill up the rules, and try to feedback the skip in PR comments, hope I can manage to do it using the asfgit credential, which I inspect it before

image

**Comment by wu-sheng** [comment]
After you verify the submodule, we could work on the rules, and notify the dev ml and committer team.

**Comment by kezhenxu94** [comment]
TODO list in this PR (or maybe following PR) for record

[x] submodule changes should trigger build
[ ] comment on PR when skipped
[ ] support manually trigger build job forcibly, in case that the condition failed to detect

@wu-sheng if I missed anything, feel free to supplement :)

**Comment by wu-sheng** [comment]
I think this is a good todo list. Follow this please

**Comment by kezhenxu94** [comment]
- [ ] list the changed file sets to trigger/skip E2E, CI, Agent Test

  - [ ] changed file sets to trigger E2E
  '^pom.xml$' \
  '^Jenkinsfile-E2E$' \
  '^apm-commons/.*$' \
  '^apm-protocol/.*$' \
  '^apm-sniffer/.*$' \
  '^apm-webapp/.*$' \
  '^oap-server/.*$' \
  '^test/e2e/.*$' \
  '^tools/conditional/.*$'

  - [ ] changed file sets to trigger agent plugin test
  '^pom.xml$' \
  '^Jenkinsfile-Agent-.*$' \
  '^apm-application-toolkit/.*$' \
  '^apm-commons/.*$' \
  '^apm-protocol/.*$' \
  '^apm-sniffer/.*$' \
  '^test/plugin/.*$' \
  '^tools/conditional/.*$'

@wu-sheng can you recheck? thanks

**Comment by wu-sheng** [comment]
Make sense to me. Please continue.

**Comment by kezhenxu94** [comment]
- [x] list the changed file sets to trigger/skip E2E, CI, Agent Test

is done

**Comment by kezhenxu94** [comment]
@wu-sheng I've double checked this PR, if it looks good to you, we can make this the first version of solutions to #3658

**Comment by kezhenxu94** [comment]
> LGTM. So, CI always run, right?

Yes, we don't set barrier for CI for now

**Comment by wu-sheng** [comment]
> Yes, we don't set barrier for CI for now

Good for me. I am going to merge this, and test on #3677, which should skip agent plugin tests :)

**Comment by wu-sheng** [comment]
Merging, and please notify the dev ml, we are doing an experimental PR testing feature for at least two weeks.

File: tools/ci/changed.sh [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    *)
+      patterns+=($1)
+      ;;
+  esac
+  shift
+done
+
+[[ ${#patterns[@]} -eq 0 ]] && echo 'No file pattern is specified, exiting' && exit 1
+
+changed_files=$(git diff --name-only origin/${ghprbTargetBranch:-master}..${ghprbActualCommit:-HEAD})
```
Comment (wu-sheng) [inline tools/ci/changed.sh]: Is this set `master` as the default target? If `ghprbTargetBranch` is missing?

Comment (kezhenxu94) [reply tools/ci/changed.sh]: `ghprbTargetBranch` is always present in Jenkins environment, it should be the former branch here 
image

Comment (kezhenxu94) [reply tools/ci/changed.sh]: default to `master` is only for testing locally, it can be safely removed in Jenkins environment

**Pull request merged**